### PR TITLE
Use DS callable API to allow hf_scheduler + ds_optimizer

### PR DESCRIPTION
**Pull request by tjruwase**
This PR:

- Used the (new) Callable api of deepspeed.initialize() to enable combining hf schedulers with deepspeed optimizers.  
- `create_scheduler` now has an optional `optimizer` arg
- Updates relevant unit test.

Blocking events: All unblocked now.

- [x] depends on deepspeed PR [1316](https://github.com/microsoft/DeepSpeed/pull/1316).
- [x] needs new deepspeed version after PR is merged and need to update the dependencies when that happens.

deepspeed: @stas00.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -782,56 +785,56 @@ def create_optimizer(self):
...
+            )
+        else:
+            optimizer = optimizer_cls(optimizer_grouped_parameters, **optimizer_kwargs)
 
         if is_sagemaker_mp_enabled():
-            self.optimizer = smp.DistributedOptimizer(self.optimizer)
+            optimizer = smp.DistributedOptimizer(optimizer)
 
-    def create_scheduler(self, num_training_steps: int):
+        return optimizer
```
Comment (stas00) [inline src/transformers/trainer.py:823]:
```suggestion
        self.optimizer = optimizer
        return optimizer
```

this breaks back-compatibility so it needs to be changed back to keep `create_optimizer` setting ` self.optimizer` internally and the same for `create_scheduler`. So the only change needed is to set the attributes beside returning to keep the functions work as before.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -782,56 +785,56 @@ def create_optimizer(self):
...
-                self.optimizer,
-                num_warmup_steps=self.args.get_warmup_steps(num_training_steps),
-                num_training_steps=num_training_steps,
-            )
+        return get_scheduler(
+            self.args.lr_scheduler_type,
+            optimizer,
+            num_warmup_steps=self.args.get_warmup_steps(num_training_steps),
+            num_training_steps=num_training_steps,
+        )
```
Comment (stas00) [inline src/transformers/trainer.py:837]:
```suggestion
        lr_scheduler = get_scheduler(
            self.args.lr_scheduler_type,
            optimizer,
            num_warmup_steps=self.args.get_warmup_steps(num_training_steps),
            num_training_steps=num_training_steps,
        )
        self.lr_scheduler = lr_scheduler
        return lr_scheduler
```

same here

**Review by stas00 — CHANGES_REQUESTED** [review]
Thank you for solving this limitation, @tjruwase 

I just proposed 2 fixes to keep things back-compatible. Please let me know if it's still OK.

File: src/transformers/trainer.py (hunk 1 above)
Comment (tjruwase) [reply src/transformers/trainer.py:823]: Thanks for the feedback. The previous code also had guards like `if self.optimizer is None:` to prevent repeated execution of the optimizer and scheduler creation logic. So I will revert back to the old code and only add the return optimizer and scheduler statements.

Comment (stas00) [reply src/transformers/trainer.py:823]: Ah yes! I guess it was used primarily internally initially so that's why it probably wasn't designed to be a user-level function.

Your suggestion to keep things as before and just add a return is probably the best. We will also run this by Sylvain once things are ready.

File: src/transformers/deepspeed.py [hunk 3]
```diff
@@ -336,28 +336,20 @@ def deepspeed_init(trainer, num_training_steps, resume_from_checkpoint=None):
...
         config["zero_allow_untested_optimizer"] = True
 
-    # DS schedulers (deepspeed/runtime/lr_schedules.py):
-    #
-    # DS name      | --lr_scheduler_type  | HF func                           | Notes
-    # -------------| ---------------------|-----------------------------------|--------------------
-    # LRRangeTest  | na                   | na                                | LRRT
-    # OneCycle     | na                   | na                                | 1CLR
-    # WarmupLR     | constant_with_warmup | get_constant_schedule_with_warmup | w/ warmup_min_lr=0
-    # WarmupDecayLR| linear               | get_linear_schedule_with_warmup   |
```
Comment (stas00) [inline src/transformers/deepspeed.py:351]: why was this table removed? I spent quite some time researching it - is it no longer correct?

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -815,10 +815,12 @@ def create_optimizer(self):
...
 
-        if is_sagemaker_mp_enabled():
-            self.optimizer = smp.DistributedOptimizer(self.optimizer)
+            if is_sagemaker_mp_enabled():
+                self.optimizer = smp.DistributedOptimizer(self.optimizer)
+
+        return self.optimizer
 
-    def create_scheduler(self, num_training_steps: int):
+    def create_scheduler(self, optimizer: torch.optim.Optimizer, num_training_steps: int):
```
Comment (stas00) [inline src/transformers/trainer.py:823]: hmm, that would break back-compat as the new arg is added before the existing one - and if someone used a positional arg to pass `num_training_steps` it now will break.

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -772,8 +772,8 @@ def create_optimizer_and_scheduler(self, num_training_steps: int):
         Trainer's init through :obj:`optimizers`, or subclass and override this method (or :obj:`create_optimizer`
         and/or :obj:`create_scheduler`) in a subclass.
         """
-        self.create_optimizer()
-        self.create_scheduler(num_training_steps)
+        self.optimizer = self.create_optimizer()
+        self.lr_scheduler = self.create_scheduler(optimizer=self.optimizer, num_training_steps=num_training_steps)
```
Comment (stas00) [inline src/transformers/trainer.py:776]: the assignment isn't needed, no?

```suggestion
        self.create_optimizer()
        self.create_scheduler(optimizer=self.optimizer, num_training_steps=num_training_steps)
```

I know that now that you changed the API to return the value, it's sort of a mixed bag.

File: src/transformers/trainer.py (hunk 4 above)
Comment (stas00) [reply src/transformers/trainer.py:823]: may be putting it last and making it optional? So by default it works as before but only if the optional `optimizer is not None` than we use that?

**Review by stas00 — COMMENTED** [review]
We will have to wait for @sgugger to finalize this as we are messing with the API. I think he's back on the 30th. in a week.

File: src/transformers/trainer.py (hunk 4 above)
Comment (tjruwase) [reply src/transformers/trainer.py:823]: Good point.

File: src/transformers/deepspeed.py (hunk 3 above)
Comment (tjruwase) [reply src/transformers/deepspeed.py:351]: Yes, it is no longer correct and I am not sure why. Moreover, it was no longer used and should be now redundant with this PR.

File: src/transformers/trainer.py (hunk 5 above)
Comment (tjruwase) [reply src/transformers/trainer.py:776]: The assignments were missing there originally as seen [here](https://github.com/huggingface/transformers/blob/2772d3e79d66925cf4adeaffd8be610f0ab177b6/src/transformers/trainer.py#L770-L771).

Comment (stas00) [reply src/transformers/trainer.py:776]: I think we are saying the same thing. 

That's why I proposed to restore how it was originally, since the proposed change is redundant.

Comment (tjruwase) [reply src/transformers/trainer.py:776]: My bad, I got confused. I thought I had removed them. Will do on the next push.

**Review by stas00 — APPROVED** [review]
Thank you for making the changes, Tunji. Looking good.

Now let's wait for the merge of your DS PR and the new release.

**Review by sgugger — APPROVED** [review]
Thanks for your PR!

I have no problem with `optimizer` becoming an optional argument of `Trainer.create_scheduler`, I am less convinced by the need to change the return type of the two methods `Trainer.create_optimizer` and `Trainer.create_scheduler`, though I don't have particularly strong feelings. I understand it's more convenient to do
```
return trainer.create_scheduler(xxx)
```
instead of
```
trainer.create_scheduler(xxx)
return trainer.lr_scheduler
```
but it's internal code.

It's just a thought, if you really prefer the change of return type, I'm fine with it.

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -813,7 +813,9 @@ def create_optimizer(self):
         if is_sagemaker_mp_enabled():
             self.optimizer = smp.DistributedOptimizer(self.optimizer)
 
-    def create_scheduler(self, num_training_steps: int):
+        return self.optimizer
+
+    def create_scheduler(self, num_training_steps: int, optimizer: torch.optim.Optimizer = None):
         """
         Setup the scheduler. The optimizer of the trainer must have been set up before this method is called.
```
Comment (stas00) [inline src/transformers/trainer.py:820]:
```suggestion
        Setup the scheduler. The optimizer of the trainer must have been set up either before this method is called or passed as an argument.
```

need to update the doc too
